packages/squiggle-lang/src/rescript/Reducer/Reducer_Dispatch/Reducer_Dispatch_BuiltInMacros.res

```diff
@@ -3,120 +3,136 @@
+  let doBindExpression = (bindingExpr: expression, statement: expression, environment) =>
     switch statement {
-    | ExpressionT.EList(list{
-        ExpressionT.EValue(EvCall("$let")),
-        ExpressionT.EValue(EvSymbol(aSymbol)),
-        expressionToReduce,
-      }) => {
-        let rNewExpressionToReduce = replaceSymbols(expressionToReduce, bindings)
+    | ExpressionT.EList(list{ExpressionT.EValue(EvCall("$let")), symbolExpr, statement}) => {
+        let rExternalBindingsValue = reduceExpression(
+          bindingExpr,
+          Belt.Map.String.fromArray([("x", ExpressionValue.EvNumber(666.))]),
+          // bindingsToHandDown,
+          environment,
+        )
 
-        let rNewValue =
-          rNewExpressionToReduce->Result.flatMap(newExpressionToReduce =>
-            reduceExpression(newExpressionToReduce, bindings)
+        rExternalBindingsValue->Result.flatMap(externalBindingsValue => {
+          let newBindings = Bindings.fromValue(externalBindingsValue)
+          let rNewStatement = Bindings.replaceSymbols(newBindings, statement)
+          rNewStatement->Result.map(newStatement =>
+            eFunction(
+              "$exportBindings",
+              list{
+                eFunction(
+                  "$setBindings",
+                  list{newBindings->Bindings.toExternalBindings->eRecord, symbolExpr, newStatement},
+                ),
+              },
+            )
           )
-
-        let rNewExpression = rNewValue->Result.map(newValue => ExpressionT.EValue(newValue))
-        rNewExpression->Result.map(newExpression =>
-          Belt.Map.String.set(bindings, aSymbol, newExpression)->ExpressionT.EBindings
-        )
+        })
+      }
+    | _ => {
+        let rExternalBindingsValue = reduceExpression(bindingExpr, bindings, environment)
+        rExternalBindingsValue->Result.flatMap(externalBindingsValue => {
+          let newBindings = Bindings.fromValue(externalBindingsValue)
+          let rNewStatement = Bindings.replaceSymbols(newBindings, statement)
+          rNewStatement
+        })
       }
-    | _ => REAssignmentExpected->Error
     }
-  }
 
-  let doExportVariableExpression = (bindings: ExpressionT.bindings) => {
-    let emptyDictionary: Js.Dict.t<ExpressionValue.expressionValue> = Js.Dict.empty()
-    let reducedBindings = bindings->Belt.Map.String.keep((_key, value) =>
-      switch value {
-      | ExpressionT.EValue(_) => true
-      | _ => false
-      }
-    )
-    let externalBindings = reducedBindings->Belt.Map.String.reduce(emptyDictionary, (
-      acc,
-      key,
-      expressionValue,
-    ) => {
-      let value = switch expressionValue {
-      | EValue(aValue) => aValue
-      | _ => EvSymbol("internal")
+  let doBlock = (exprs: list<expression>, _bindings: ExpressionT.bindings, _environment): result<
+    expression,
+    errorValue,
+  > => {
+    let exprsArray = Belt.List.toArray(exprs)
+    let maxIndex = Js.Array2.length(exprsArray) - 1
```

Comment:
Any reason not to use ``E`` now, like, ``E.L.toArray``, and ``E.A.maxIndex``? (If the function isn't there yet, feel free to add)

Reply:
Good suggestion... The code needs a pass for E